### ISSUE TITLE
Fully reproducible repo2docker builds

### ISSUE THREAD

**Issue by weiji14** [CryoInTheCloud/hub-image#11]
Ensure that people can reproduce the cryocloud software environment locally and in the future.

Some actionable steps:
- [ ] Tag versions of the docker image that has been deployed to the CryoCloud JupyterHub
- [ ] Use lockfiles and or version pinning for dependencies installed via conda/pip/apt
- [ ] Ensure cross platform compatibility on Linux/macOS/Windows

Relevant discussions previously:

        @weiji14 - absolutely!! My take on this issue is that we should take a "two-level" approach:

* Once we settle on a complete list of packages for the base `environment.yml` file, then we run a script @yuvipanda wrote that will write in the current conda-forge package versions in it. But that's _not_ a conda lock file, b/c it remains platform independent as it won't have build hashes in it. This makes it easy for users to replicate the environment locally across OSes, which I think is a very important feature (not everything can be done locally, but it's very valuable for users to at least have a reproducible offline workflow that matches the cloud service as much as possible).
* _Then_, we generate a `conda-lock.yml` file to pin the docker images with exact builds. We tag the image with a version number at that point. 

Regarding versioning, my suggestion would be to use an `2022.X.Y` scheme, where the "major" number `X` changes when new packages are added, whereas the "minor" `Y` changes when any package version or even build is updated. This will let people say "This project needs the cryohub environment `2022.3.4`" with that being a pretty unambiguous reference.  

As for the apt packages, I think that's a bit less of a concern, since they change more slowly and we can ensure that a given `YYYY` series never changes the underlying Ubuntu base image, so only minor version fixes will go in. I don't know of the equivalent of a conda-lock for apt packages, but that's probably just my ignorance.

_Originally posted by @fperez in https://github.com/CryoInTheCloud/CryoCloudWebsite/issues/1#issuecomment-1326822490_
      

        Ideally, we would use this `conda-linux-64.lock` to build the docker image for full reproducibility, but there are some PyPI-only packages right now that are not on conda-forge.

@yuvipanda, have you thought about creating conda-forge packages for https://github.com/yuvipanda/jupyter-desktop-server and https://github.com/yuvipanda/jupyter-syncthing-proxy?

_Originally posted by @weiji14 in https://github.com/CryoInTheCloud/hub-image/pull/9#discussion_r1034082681_

**Comment by fperez** [comment]
Awesome - many thanks for opening this here so we track progress!

@tsnow03 we should mention this on Friday so the community users are aware of the longer-term view regarding these points, even if not all the pieces will be in place by then.

**Comment by yuvipanda** [comment]
Also, I *think* right now repo2docker doesn't actually respect the lockfile at all, so I don't think it's used. The 'right' way to fix this is to add lockfile support to repo2docker. Until then, I suggest:

1. renaming the environment.yml to something else, so it doesn't trigger the repo2docker behavior
2. Adding a `postBuild` bash file that installs the environment from the lockfile

**Comment by weiji14** [comment]
> Also, I _think_ right now repo2docker doesn't actually respect the lockfile at all, so I don't think it's used. The 'right' way to fix this is to add lockfile support to repo2docker.

Yep, that is being tracked by https://github.com/jupyterhub/repo2docker/issues/1157.

> Until then, I suggest:
> 
> 1. renaming the environment.yml to something else, so it doesn't trigger the repo2docker behavior
> 
> 2. Adding a `postBuild` bash file that installs the environment from the lockfile

Ok, then we will need to use the [unified `conda-lock.yml`](https://conda-incubator.github.io/conda-lock/output) file rather than the current [explicit lockfile `conda-linux-64.lock`](https://conda-incubator.github.io/conda-lock/output/#explicit-lockfile) which doesn't handle pip-only dependencies. Let me see if I can get to working using [`postBuild`](https://repo2docker.readthedocs.io/en/latest/config_files.html#postbuild-run-code-after-installing-the-environment).

**Comment by weiji14** [comment]
> @yuvipanda, have you thought about creating conda-forge packages for https://github.com/yuvipanda/jupyter-desktop-server and https://github.com/yuvipanda/jupyter-syncthing-proxy?

Turns out we really do need these two packages to be on conda-forge, otherwise conda-lock won't work, see https://github.com/CryoInTheCloud/hub-image/pull/14#issuecomment-1331452732. So I've started Pull Requests at https://github.com/conda-forge/staged-recipes/pull/21368 and https://github.com/conda-forge/staged-recipes/pull/21369 to add those to conda-forge :slightly_smiling_face:
